node/rpc/src/lib.rs

```diff
@@ -51,114 +76,421 @@ where
+                                                        chain_notify_count,
+                                                        e.message()
+                                                    ),
+                                                )
+                                                .await
+                                                .unwrap_or_else(|e| {
+                                                    error!(
+                                                        "error {:?} on socket {:?}",
+                                                        e.message(),
+                                                        addr
+                                                    )
+                                                });
+                                            })
+                                            .await
+                                            .unwrap_or_else(|_| {
+                                                error!("error sending on socket {:?}", addr)
+                                            });
+
+                                        handle_events_out
+                                            .write()
+                                            .await
+                                            .publish(EventsPayload::TaskCancel(
+                                                chain_notify_count,
+                                                (),
+                                            ))
+                                            .await;
+                                    } else {
+                                        handle_events_out
+                                            .write()
+                                            .await
+                                            .publish(EventsPayload::TaskCancel(
+                                                chain_notify_count,
+                                                (),
+                                            ))
+                                            .await
+                                    }
+                                });
+                            }
+                            Err(e) => send_error(1, ws_sender.clone(), e.to_string())
+                                .await
+                                .unwrap_or_else(|e| {
+                                    error!("error {:?} on socket {:?}", e.message(), addr)
+                                }),
+                        }
+                    }
+                    Err(e) => send_error(2, ws_sender.clone(), e.to_string())
+                        .await
+                        .unwrap_or_else(|e| error!("error {:?} on socket {:?}", e.message(), addr)),
+                };
+            }
+        } else {
+            warn!("web socket connection failed at {:}", addr)
+        }
+    })
+}
+
+async fn send_error(
+    code: i64,
+    ws_sender: Arc<RwLock<WsSink>>,
```

Comment:
just a thought, can this not just be `&mut WsSink` and just await a write lock on the function call so you don't need to clone the arc when you are awaiting the response anyway?

Reply:
The downside of that is the lock will be obtained for the whole lifetime of the function rather than just the single line that needs the send but I can change

Reply:
Ah ya, I guess I meant a ref to the RwLock, to avoid the arc clone